src/usb.ts

```diff
@@ -64,29 +67,29 @@ export class USB extends EventDispatcher {
         });
 
         adapter.addListener(USBAdapter.EVENT_DEVICE_DISCONNECT, handle => {
-            const allowedDevice = this.allowedDevices.find(allowedDevices => allowedDevices._handle === handle);
+            const device = this.deviceByHandle.get(handle);
 
-            if (allowedDevice) {
-                this.emit(USB.EVENT_DEVICE_DISCONNECT, allowedDevice);
+            if (device) {
+                const key = deviceToKey(device);
 
-                // Delete device from cache
-                const index = this.allowedDevices.indexOf(allowedDevice);
-                this.allowedDevices.splice(index, 1);
+                this.emit(USB.EVENT_DEVICE_DISCONNECT, device);
+                this.connectedDevices.delete(key);
+                this.deviceByHandle.delete(handle);
             }
         });
     }
 
     private replaceAllowedDevice(device: USBDevice): boolean {
-        for (const i in this.allowedDevices) {
-            if (this.allowedDevices[i].productId === device.productId
-                && this.allowedDevices[i].vendorId === device.vendorId
-                && this.allowedDevices[i].serialNumber === device.serialNumber) {
-                this.allowedDevices[i] = device;
-                return true;
-            }
+        const key = deviceToKey(device);
+
+        if (!this.allowedDevices.has(key)) {
+            return false;
         }
 
```

Comment:
I was going to add another assert here, but decided against it as it should not happen. But will still write here:

```typescript
        if (this.connectedDevices.has(key)) {
            const oldHandle = this.connectedDevices.get(key)._handle;

            if (oldHandle && oldHandle !== device._handle) {
                // Assertion error.
                // this should not happen, because disconnected
                // devices will be cleaned up from device handle map
                // on disconnection.
                throw new Error("Device handle clean up error.");
            }
        }
```